### PR TITLE
fix: unify scores

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
Reference pr: https://github.com/jina-ai/jina/pull/2942

### CODE REVIEW

File: jinahub/indexers/searcher/FaissSearcher/__init__.py [hunk 1]
```diff
@@ -220,7 +224,10 @@ def search(
             for m_info in zip(*matches):
                 idx, dist = m_info
                 match = Document(id=self._ids[idx], embedding=self._vecs[idx])
-                match.scores['distance'] = dist
+                if self.reverse_score:
+                    match.scores['similarity'] = 1 / (1 + dist)
+                else:
+                    match.scores['distance'] = dist
```
Comment (numb3r3) [inline jinahub/indexers/searcher/FaissSearcher/__init__.py:230]: same

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: jinahub/indexers/searcher/AnnoySearcher/__init__.py [hunk 2]
```diff
@@ -84,7 +87,10 @@ def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
             )
             for idx, dist in zip(indices, dists):
                 match = Document(id=self._ids[idx], embedding=self._vecs[idx])
-                match.scores['distance'] = 1 / (1 + dist)
+                if self.reverse_score:
+                    match.scores['similarity'] = 1 / (1 + dist)
+                else:
+                    match.scores['distance'] = dist
```
Comment (JoanFM) [inline jinahub/indexers/searcher/AnnoySearcher/__init__.py:93]: the name of the key should relate to the metric

File: jinahub/indexers/searcher/AnnoySearcher/__init__.py [hunk 3]
```diff
@@ -84,7 +87,10 @@ def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
             )
             for idx, dist in zip(indices, dists):
                 match = Document(id=self._ids[idx], embedding=self._vecs[idx])
-                match.scores['distance'] = 1 / (1 + dist)
+                if self.reverse_score:
+                    match.scores['similarity'] = 1 / (1 + dist)
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply jinahub/indexers/searcher/AnnoySearcher/__init__.py:91]: This conversion is not good for cosine or dot product, in that case the conversion should simply be 1 - dist

**Review by JoanFM — CHANGES_REQUESTED** [review]
You have to be very careful. Some packages return similarity already depending on the type of metric asked for

File: jinahub/indexers/searcher/AnnoySearcher/__init__.py (hunk 2 above)
Comment (Yongxuanzhang) [reply jinahub/indexers/searcher/AnnoySearcher/__init__.py:93]: In ranker when we want to extract the scores values, should I specify the keys like 'distance' 'similarity'? Now in customized ranker I have something like this, if I have different names, How can I get the values?

`chunk_match_list.sort(key=lambda m: -m.scores['distance'].value)`

**Comment by Yongxuanzhang** [comment]
> You have to be very careful. Some packages return similarity already depending on the type of metric asked for

I see, I will check each metric for in our searchers

File: jinahub/indexers/searcher/AnnoySearcher/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yongxuanzhang) [reply jinahub/indexers/searcher/AnnoySearcher/__init__.py]: done

File: jinahub/indexers/searcher/AnnoySearcher/__init__.py (hunk 2 above)
Comment (Yongxuanzhang) [reply jinahub/indexers/searcher/AnnoySearcher/__init__.py]: done

**Review by JoanFM — CHANGES_REQUESTED** [review]
specially for faiss buf also fof the others, have a padametrized test that shows that for every supported metric, when distance is required the scores are sorted in an increasing manner, and when similarity they are ordered in a decreasing one

**Comment by Yongxuanzhang** [comment]
> specially for faiss buf also fof the others, have a padametrized test that shows that for every supported metric, when distance is required the scores are sorted in an increasing manner, and when similarity they are ordered in a decreasing one

yes I have added tests like this I think @JoanFM

File: jinahub/indexers/searcher/AnnoySearcher/__init__.py [hunk 4]
```diff
@@ -84,7 +87,18 @@ def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
...
+                        match.scores[self.metric] = 1 - dist
+                    else:
+                        match.scores[self.metric] = dist
+                else:
+                    if self.metric == 'angular' or self.metric == 'hamming':
+                        match.scores[self.metric] = 1 - dist
+                    elif self.metric == 'euclidean' or self.metric == 'manhattan':
+                        match.scores[self.metric] = 1 / (1 + dist)
+                    else:
+                        match.scores[self.metric] = dist
```
Comment (JoanFM) [inline jinahub/indexers/searcher/AnnoySearcher/__init__.py:101]: it is weird to say `is_distance` False and set `score` to `distance`

File: jinahub/indexers/searcher/FaissSearcher/tests/test_faissindexer.py [hunk 5]
```diff
@@ -91,11 +91,43 @@ def test_faiss_indexer(metas, tmpdir_dump):
...
+                d.matches[0].scores[indexer.metric].value
+                >= d.matches[1].scores[indexer.metric].value
         )
 
 
+@pytest.mark.parametrize(['metric', 'is_distance'],
+                         [('l2', True), ('inner_product', True),
+                          ('l2', False), ('inner_product', False)])
+def test_faiss_metric(metas, tmpdir_dump, metric, is_distance):
+    print(f'dump path = {tmpdir_dump}')
```
Comment (JoanFM) [inline jinahub/indexers/searcher/FaissSearcher/tests/test_faissindexer.py:103]: remove print

File: jinahub/indexers/searcher/HnswlibSearcher/tests/test_hnswlibsearcher.py [hunk 6]
```diff
@@ -29,6 +29,23 @@ def metas(tmpdir):
...
+                         [('l2', True), ('ip', True), ('cosine', True),
+                          ('l2', False), ('ip', False), ('cosine', False)])
+def test_metric(tmpdir, metric, is_distance):
+    metas = {'workspace': str(tmpdir), 'name': 'searcher', 'pea_id': 0, 'replica_id': 0}
+
+    indexer = HnswlibSearcher(dump_path=DUMP_PATH, top_k=TOP_K, metas=metas, metric=metric, is_distance=is_distance)
+    docs = DocumentArray([Document(embedding=np.random.random(7))])
+    indexer.search(docs, {})
+
+    for i in range(len(docs[0].matches) - 1):
```
Comment (JoanFM) [inline jinahub/indexers/searcher/HnswlibSearcher/tests/test_hnswlibsearcher.py:42]: assert length of matches before. To make sure the `for` actually happens

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/indexers/searcher/HnswlibSearcher/tests/test_hnswlibsearcher.py (hunk 6 above)
Comment (Yongxuanzhang) [reply jinahub/indexers/searcher/HnswlibSearcher/tests/test_hnswlibsearcher.py:42]: ah yes, sure

File: jinahub/indexers/searcher/AnnoySearcher/__init__.py (hunk 4 above)
Comment (Yongxuanzhang) [reply jinahub/indexers/searcher/AnnoySearcher/__init__.py:101]: This line is only for `dot`, the inner product metric. The annoy already output `dot` as a similarity score.
So the `is_distance ==True` leads to `1-dist`, `is_distance ==False` leads to original `dist`.
Should I add some comments above these code?

Comment (Yongxuanzhang) [reply jinahub/indexers/searcher/AnnoySearcher/__init__.py]: I refactored those check

File: jinahub/indexers/searcher/NumpySearcher/tests/test_query.py [hunk 7]
```diff
@@ -20,6 +23,30 @@ def test_query_vector(tmpdir):
...
+                          ('cosine', False), ('euclidean', False)])
+def test_metric(tmpdir, metric, is_distance):
+    runtime = {
+        'workspace': str(tmpdir),
+        'name': 'searcher',
+        'pea_id': 0,
+        'replica_id': 0,
+    }
+
+    indexer = NumpySearcher(dump_path='tests/dump1', default_top_k=TOP_K, runtime_args=runtime, metric=metric,
```
Comment (JoanFM) [inline jinahub/indexers/searcher/NumpySearcher/tests/test_query.py:37]: can u make this path relative to cur_dir?

**Review by JoanFM — APPROVED** [review]
LGTM

File: jinahub/indexers/searcher/NumpySearcher/tests/test_query.py (hunk 7 above)
Comment (Yongxuanzhang) [reply jinahub/indexers/searcher/NumpySearcher/tests/test_query.py]: how about now?

**Pull request merged**